File: widgetssdk/src/main/java/com/glia/widgets/chat/adapter/holder/OperatorStatusViewHolder.java
package com.glia.widgets.chat.adapter.holder;

import android.content.Context;
import android.graphics.Color;
import android.graphics.Typeface;
import android.view.View;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import androidx.core.content.res.ResourcesCompat;
import androidx.recyclerview.widget.RecyclerView;

import com.glia.widgets.R;
import com.glia.widgets.UiTheme;
import com.glia.widgets.chat.model.history.OperatorStatusItem;
import com.glia.widgets.view.OperatorStatusView;
import com.glia.widgets.view.configuration.chat.ChatStyle;

public class OperatorStatusViewHolder extends RecyclerView.ViewHolder {
    private final OperatorStatusView statusPictureView;
    private final TextView chatStartingHeadingView;
    private final TextView chatStartingCaptionView;
    private final TextView chatStartedNameView;
    private final TextView chatStartedCaptionView;
    private final Context context;
    private final ChatStyle chatStyle;

    public OperatorStatusViewHolder(@NonNull View itemView, UiTheme uiTheme, ChatStyle chatStyle) {
        super(itemView);
        this.chatStyle = chatStyle;
        this.statusPictureView = itemView.findViewById(R.id.status_picture_view);
        this.chatStartingHeadingView = itemView.findViewById(R.id.chat_starting_heading_view);
        this.chatStartingCaptionView = itemView.findViewById(R.id.chat_starting_caption_view);
        this.chatStartedNameView = itemView.findViewById(R.id.chat_started_name_view);
        this.chatStartedCaptionView = itemView.findViewById(R.id.chat_started_caption_view);

        context = itemView.getContext();

        statusPictureView.setTheme(uiTheme , chatStyle);

        setStartedHeadingTextColor(uiTheme);
        setStartedCaptionTextColor(uiTheme);

        if (uiTheme.getFontRes() != null) {
            Typeface fontFamily = ResourcesCompat.getFont(context, uiTheme.getFontRes());
            chatStartingHeadingView.setTypeface(fontFamily, Typeface.BOLD);
            chatStartingCaptionView.setTypeface(fontFamily);
            chatStartedNameView.setTypeface(fontFamily, Typeface.BOLD);
            chatStartedCaptionView.setTypeface(fontFamily);
        }

        chatStartingCaptionView.setText(chatStyle.welcomeView.descriptionValue);
        chatStartingCaptionView.setTextColor(Color.parseColor(chatStyle.welcomeView.description.foregroundColor));
        chatStartingCaptionView.setTextSize(chatStyle.welcomeView.description.getTextSize());
    }

    private void setStartedHeadingTextColor(UiTheme uiTheme) {
        chatStartedNameView.setTextColor(
                ContextCompat.getColor(
                        context,
                        uiTheme.getGliaChatStartedHeadingTextColor()
                )
        );
    }

    private void setStartedCaptionTextColor(UiTheme uiTheme) {
        chatStartedCaptionView.setTextColor(
                ContextCompat.getColor(
                        context,
                        uiTheme.getGliaChatStartedCaptionTextColor()
                )
        );
    }

    public void bind(OperatorStatusItem item) {
        chatStartingHeadingView.setText(chatStyle.welcomeView.titleValue);
        chatStartingHeadingView.setTextColor(Color.parseColor(chatStyle.welcomeView.title.foregroundColor));
        chatStartingHeadingView.setTextSize(chatStyle.welcomeView.title.getTextSize());
        if (item.getStatus() == OperatorStatusItem.Status.IN_QUEUE) {
            statusPictureView.showPlaceholder();
            chatStartingHeadingView.setVisibility(View.VISIBLE);
            chatStartingCaptionView.setVisibility(View.VISIBLE);
            chatStartedNameView.setVisibility(View.GONE);
            chatStartedCaptionView.setVisibility(View.GONE);
            itemView.setContentDescription(context.getString(R.string.glia_chat_in_queue_message_content_description, item.getCompanyName()));
        } else if (item.getStatus() == OperatorStatusItem.Status.OPERATOR_CONNECTED) {
            if (item.getProfileImgUrl() != null) {
                statusPictureView.showProfileImage(item.getProfileImgUrl());
            } else {
                statusPictureView.showPlaceholder();
            }
            chatStartedNameView.setText(item.getOperatorName());
            chatStartedCaptionView.setText(context.getString(R.string.glia_chat_operator_has_joined, item.getOperatorName()));

            chatStartingHeadingView.setVisibility(View.GONE);
            chatStartingCaptionView.setVisibility(View.GONE);
            chatStartedNameView.setVisibility(View.VISIBLE);
            chatStartedCaptionView.setVisibility(View.VISIBLE);

            itemView.setContentDescription(context.getString(R.string.glia_chat_operator_has_joined_content_description, item.getOperatorName()));
        } else if (item.getStatus() == OperatorStatusItem.Status.JOINED) {
            chatStartedNameView.setText(item.getOperatorName());
            chatStartedCaptionView.setText(context.getString(R.string.glia_chat_operator_has_joined, item.getOperatorName()));
            chatStartingHeadingView.setVisibility(View.GONE);
            chatStartingCaptionView.setVisibility(View.GONE);
            chatStartedNameView.setVisibility(View.VISIBLE);
            chatStartedCaptionView.setVisibility(View.VISIBLE);
            itemView.setContentDescription(context.getString(R.string.glia_chat_operator_has_joined_content_description, item.getOperatorName()));
        } else if (item.getStatus() == OperatorStatusItem.Status.TRANSFERRING) {
            statusPictureView.showPlaceholder();
            chatStartingHeadingView.setVisibility(View.VISIBLE);
            chatStartingCaptionView.setVisibility(View.VISIBLE);
            chatStartedNameView.setVisibility(View.VISIBLE);
            chatStartedNameView.setText(context.getString(R.string.glia_chat_visitor_status_transferring));
            chatStartedCaptionView.setVisibility(View.GONE);
        }
        statusPictureView.setVisibility(
                isShowStatusPictureView(item.getStatus()) ? View.VISIBLE : View.GONE
        );
        statusPictureView.setShowRippleAnimation(isShowStatusViewRippleAnimation(item));
    }

    private boolean isShowStatusPictureView(OperatorStatusItem.Status status) {
        return status != OperatorStatusItem.Status.JOINED;
    }

    private boolean isShowStatusViewRippleAnimation(OperatorStatusItem item) {
        return item.getStatus() == OperatorStatusItem.Status.IN_QUEUE ||
                item.getStatus() == OperatorStatusItem.Status.TRANSFERRING;
    }
}

File: widgetssdk/src/main/java/com/glia/widgets/chat/adapter/holder/OperatorStatusViewHolder.kt
package com.glia.widgets.chat.adapter.holder

import android.graphics.Typeface
import android.view.View
import android.widget.TextView
import androidx.core.view.isVisible
import androidx.recyclerview.widget.RecyclerView
import com.glia.widgets.R
import com.glia.widgets.chat.model.OperatorStatusItem
import com.glia.widgets.databinding.ChatOperatorStatusLayoutBinding
import com.glia.widgets.di.Dependencies
import com.glia.widgets.helper.applyGliaThemeFont
import com.glia.widgets.helper.setLocaleContentDescription
import com.glia.widgets.helper.setLocaleText
import com.glia.widgets.locale.StringKey
import com.glia.widgets.locale.StringKeyPair
import com.glia.widgets.view.OperatorStatusView
import com.glia.widgets.view.unifiedui.applyTextTheme
import com.glia.widgets.view.unifiedui.theme.chat.EngagementStateTheme
import com.glia.widgets.view.unifiedui.theme.chat.EngagementStatesTheme

internal class OperatorStatusViewHolder(
    binding: ChatOperatorStatusLayoutBinding
) : RecyclerView.ViewHolder(binding.root) {
    private val statusPictureView: OperatorStatusView by lazy { binding.statusPictureView }
    private val chatStartingHeadingView: TextView by lazy { binding.chatStartingHeadingView }
    private val chatStartingCaptionView: TextView by lazy { binding.chatStartingCaptionView }
    private val chatStartedNameView: TextView by lazy { binding.chatStartedNameView }
    private val chatStartedCaptionView: TextView by lazy { binding.chatStartedCaptionView }

    private val engagementStatesTheme: EngagementStatesTheme? by lazy {
        Dependencies.gliaThemeManager.theme?.chatTheme?.connect
    }

    init {
        applyBaseConfig()
        setBaseStrings()
    }

    private fun setBaseStrings() {
        chatStartingCaptionView.setLocaleText(R.string.engagement_connection_screen_message)
    }

    /**
     * The four text colours come straight from the layout's `?attr/gliaChatStart*` values; only the
     * theme typeface still has to be applied here, since a `textAppearance` would otherwise win.
     */
    private fun applyBaseConfig() {
        itemView.context.applyGliaThemeFont(chatStartingHeadingView, chatStartedNameView, style = Typeface.BOLD)
        itemView.context.applyGliaThemeFont(chatStartingCaptionView, chatStartedCaptionView)
        engagementStatesTheme?.operator.also(statusPictureView::applyOperatorTheme)
    }

    fun bind(item: OperatorStatusItem) {
        chatStartingHeadingView.setLocaleText(R.string.general_company_name)
        when (item) {
            is OperatorStatusItem.Connected -> applyConnectedState(item.operatorName, item.profileImgUrl)
            is OperatorStatusItem.InQueue -> applyInQueueState()
            is OperatorStatusItem.Joined -> applyConnectedState(item.operatorName, item.profileImgUrl)
            is OperatorStatusItem.Transferring -> applyTransferringState()
        }
        statusPictureView.isVisible = true
        statusPictureView.setShowRippleAnimation(isShowStatusViewRippleAnimation(item))
    }

    private fun applyInQueueState() {
        statusPictureView.showPlaceholder()
        applyChatStartingViewsVisibility()
        applyChatStartedViewsVisibility(false)
        itemView.setLocaleContentDescription(
            R.string.android_chat_queue_message_accessibility_label,
            StringKeyPair(StringKey.COMPANY_NAME, "")
        )

        engagementStatesTheme?.queue.also(::applyEngagementState)
    }

    private fun applyConnectedState(operatorName: String, profileImgUrl: String?) {
        profileImgUrl?.let { statusPictureView.showProfileImage(it) } ?: statusPictureView.showPlaceholder()

        applyChatStartingViewsVisibility(false)
        applyChatStartedViewsVisibility()

        chatStartedNameView.text = operatorName
        chatStartedCaptionView.setLocaleText(
            R.string.chat_operator_joined_system_message,
            StringKeyPair(StringKey.OPERATOR_NAME, operatorName)
        )
        itemView.setLocaleContentDescription(
            R.string.chat_operator_joined_system_message,
            StringKeyPair(StringKey.OPERATOR_NAME, operatorName)
        )

        engagementStatesTheme?.connected.also(::applyEngagementState)
    }

    private fun applyTransferringState() {
        statusPictureView.showPlaceholder()
        chatStartingCaptionView.isVisible = true
        chatStartedNameView.isVisible = true
        chatStartedCaptionView.isVisible = false
        chatStartingHeadingView.visibility = View.INVISIBLE

        chatStartedNameView.setLocaleText(R.string.engagement_queue_transferring)

        engagementStatesTheme?.transferring.also(::applyEngagementState)
    }

    private fun applyChatStartingViewsVisibility(visible: Boolean = true) {
        chatStartingCaptionView.isVisible = visible
        chatStartingHeadingView.isVisible = visible
    }

    private fun applyChatStartedViewsVisibility(visible: Boolean = true) {
        chatStartedNameView.isVisible = visible
        chatStartedCaptionView.isVisible = visible
    }

    private fun applyEngagementState(engagementStateTheme: EngagementStateTheme?) {
        engagementStateTheme?.description?.also {
            chatStartingCaptionView.applyTextTheme(it)
            chatStartedCaptionView.applyTextTheme(it)
        }
        engagementStateTheme?.title?.also {
            chatStartingHeadingView.applyTextTheme(it)
            chatStartedNameView.applyTextTheme(it)
        }
    }

    private fun isShowStatusViewRippleAnimation(item: OperatorStatusItem): Boolean = when (item) {
        is OperatorStatusItem.InQueue, is OperatorStatusItem.Transferring -> true
        else -> false
    }
}
